db/db_with_timestamp_basic_test.cc

```diff
@@ -590,8 +590,130 @@ TEST_F(DBBasicTestWithTimestamp, CompactDeletionWithTimestampMarkerToBottom) {
+    } else {
+      ASSERT_OK(db_->Put(write_opts, cfh, test_data_[index].key,
+                         test_data_[index].value));
+    }
+  }
+
+  void AssertVisibility(int ts, SequenceNumber seq, Status statuses[]) {
+    for (int i = 0; i < kTestDataSize; i++) {
+      if (test_data_[i].seq_num <= seq && test_data_[i].ts <= ts) {
+        ASSERT_OK(statuses[i]);
+      } else {
+        ASSERT_TRUE(statuses[i].IsNotFound());
+      }
+    }
+  }
+
+  std::vector<Slice> GetKeys() {
+    std::vector<Slice> ret(kTestDataSize);
+    for (int i = 0; i < kTestDataSize; i++) {
+      ret[i] = test_data_[i].key;
+    }
+    return ret;
+  }
+
+  void VerifyDefaultCF(int ts, const Snapshot* snap = nullptr) {
+    ReadOptions read_opts;
+    std::string read_ts = Timestamp(ts, 0);
+    Slice read_ts_slice = read_ts;
+    read_opts.timestamp = &read_ts_slice;
+    read_opts.snapshot = snap;
+
+    ColumnFamilyHandle* cfh = db_->DefaultColumnFamily();
+    std::vector<ColumnFamilyHandle*> cfs(kTestDataSize, cfh);
+    SequenceNumber seq =
+        snap ? snap->GetSequenceNumber() : kMaxSequenceNumber - 1;
+
+    // There're several MultiGet interfaces with not exactly the same
+    // implementations, query data with all of them.
+    auto keys = GetKeys();
+    std::vector<std::string> values;
+    auto s1 = db_->MultiGet(read_opts, cfs, keys, &values);
+    AssertVisibility(ts, seq, s1.data());
+
+    auto s2 = db_->MultiGet(read_opts, keys, &values);
+    AssertVisibility(ts, seq, s2.data());
+
+    std::vector<std::string> timestamps;
+    auto s3 = db_->MultiGet(read_opts, cfs, keys, &values, &timestamps);
+    AssertVisibility(ts, seq, s3.data());
+
+    auto s4 = db_->MultiGet(read_opts, keys, &values, &timestamps);
+    AssertVisibility(ts, seq, s4.data());
+
+    PinnableSlice values_ps5[kTestDataSize];
+    Status s5[kTestDataSize];
+    db_->MultiGet(read_opts, cfh, kTestDataSize, keys.data(), &values_ps5[0],
+                  &s5[0]);
+    AssertVisibility(ts, seq, s5);
+
+    PinnableSlice values_ps6[kTestDataSize];
+    Status s6[kTestDataSize];
+    std::string timestamps_array[kTestDataSize];
+    db_->MultiGet(read_opts, cfh, kTestDataSize, keys.data(), &values_ps6[0],
+                  &timestamps_array[0], &s6[0]);
+    AssertVisibility(ts, seq, s6);
+
+    PinnableSlice values_ps7[kTestDataSize];
+    Status s7[kTestDataSize];
+    db_->MultiGet(read_opts, kTestDataSize, cfs.data(), keys.data(),
+                  &values_ps7[0], &s7[0]);
+    AssertVisibility(ts, seq, s7);
+
+    PinnableSlice values_ps8[kTestDataSize];
+    Status s8[kTestDataSize];
+    db_->MultiGet(read_opts, kTestDataSize, cfs.data(), keys.data(),
+                  &values_ps8[0], &timestamps_array[0], &s8[0]);
+    AssertVisibility(ts, seq, s8);
+  }
+
+  void VerifyDefaultCF(const Snapshot* snap = nullptr) {
+    for (int i = 0; i <= kTestDataSize; i++) {
+      VerifyDefaultCF(i, snap);
+    }
+  }
 };
+constexpr int DataVisibilityTest::kTestDataSize;
```

Comment:
Nit: is this needed?

Reply:
As it's a static variable, it has to be declared somewhere.

Reply:
hmm, for constexpr pod type, it seems unnecessary (verified with g++ 11).
```
#include <cstdio>

class TestClass {
 public:
  constexpr static int kConst =1;
};

class TestClass1 {
 public:
  const static int kConst;
};

// uncomment me
// const int TestClass1::kConst = 1;

int main() {
  fprintf(stdout, "%d %d\n", TestClass::kConst, TestClass1::kConst);
  return 0;
}
```